pep-0612.rst

```diff
@@ -0,0 +1,647 @@
+       pass
+
+
+Because the default kind of parameter in Python (\ ``(x: int)``\ ) may be
+addressed both positionally and through its name, two valid invocations of
+a ``(*args: P.args, **kwargs: P.kwargs)`` function may give
+different partitions of the same set of parameters. Therefore we need to make
+sure that these special types are only brought into the world together, and are
+used together, so that our usage is valid for all possible partitions.
+
+With those requirements met, we can now take advantage of the unique properties
+afforded to us by this set up:
+
+
+* Inside the function, ``args`` has the type ``P.args``\ , not
+  ``Tuple[P.args, ...]`` as would be with a normal annotation
+  (and likewise with the ``**kwargs``\ )
+* A function of type ``Callable[P, R]`` can be called with ``(*args, **kwargs)``
+  if and only if ``args`` has the type ``P.args`` and ``kwargs`` has the type
+  ``P.kwargs``\ , and that those types both originated from the same function
+  declaration.
+* A function declared as ``def inner(*args: P.args, **kwargs: P.kwargs) -> X``
+  has type ``Callable[P, X]``.
+
+With these three properties, we now have the ability to fully type check
+parameter preserving decorators.
+
+.. code-block::
+
+   def decorator(f: Callable[P, int]) -> Callable[P, None]:
+
+     def foo(*args: P.args, **kwargs: P.kwargs) -> None:
+
+       f(*args, **kwargs)    # Accepted, should resolve to int
+
+       f(*kwargs, **args)    # Rejected
+
+       f(1, *args, **kwargs) # Rejected
+
+     return foo              # Accepted
+
+To extend this to include ``Concatenate``, we declare the following properties:
+
+* A function of type ``Callable[Concatenate[A, B, P], R]`` can only be
+  called with ``(a, b, *args, **kwargs)`` when ``args`` and ``kwargs`` are the
+  respective components of ``P``, ``a`` is of type ``A`` and ``b`` is of
+  type ``B``.
+* A function declared as
+  ``def inner(a: A, b: B, *args: P.args, **kwargs: P.kwargs) -> R``
+  has type ``Callable[Concatenate[A, B, P], R]``. Placing keyword-only
+  parameters beterrn the ``*args`` and ``**kwargs`` is forbidden.
+
+.. code-block::
+
+   def add(f: Callable[P, int]) -> Callable[Concatenate[str, P], None]:
+
+     def foo(s: str, *args: P.args, **kwargs: P.kwargs) -> None:  # Accepted
+       pass
+
+     def bar(*args: P.args, s: str, **kwargs: P.kwargs) -> None:  # Rejected
+       pass
+
+     return foo                                                   # Accepted
+
+
+   def remove(x: Callable[Concatenate[int, P], int]) -> Callable[P, None]:
+
+     def foo(*args: P.args, **kwargs: P.kwargs) -> None:
+       f(1, *args, **kwargs) # Accepted
+
+       f(*args, 1, **kwargs) # Rejected
+
+       f(*args, **kwargs)    # Rejected
+
+     return foo
+
+We also consider functions with a ``ParamSpec`` included in one of the preceding
+anonymous parameters to have the variable "already in scope" for the purposes of
```

Comment:
Not clear why `f` is an "anonymous parameter". Could you move this to the end of this section, so that we know that `foo(x: int, *args: P.args, ...)` is actually equivalent to `foo(__x: int, *args: P.args, ...)`?